### PR TITLE
Support `df.reset_index` and `series.reset_index`

### LINKED ISSUE #908: Implement `df.reset_index()` and `series.reset_index()`
**Is your feature request related to a problem? Please describe.**
Support `df.reset_index()` and `series.reset_index()`, Series will return a DataFrame when `drop=False`(the default).

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Support `reset_index` for DataFrame and Series.

## Related issue number

Resolves #908

### CODE REVIEW

File: mars/dataframe/base/reset_index.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        columns_splits = list(in_df.nsplits[1])
+        columns_splits[0] += added_columns_num
+        nsplits = (in_df.nsplits[0], tuple(columns_splits))
+        return new_op.new_dataframes(op.inputs, out_df.shape, nsplits=nsplits,
+                                     chunks=out_chunks, dtypes=out_df.dtypes,
+                                     index_value=out_df.index_value, columns_value=out_df.columns_value)
+
+    @classmethod
+    def tile(cls, op):
+        if isinstance(op.inputs[0], DATAFRAME_TYPE):
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: Better to judge by `output_type`.

File: mars/dataframe/base/reset_index.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            return self.new_dataframe([a], shape=shape, index_value=parse_index(empty_df.index),
+                                      columns_value=parse_index(empty_df.columns),
+                                      dtypes=empty_df.dtypes)
+
+    def _call_dataframe(self, a):
+        if self.drop:
+            shape = a.shape
+            columns_value = a.columns_value
+            dtypes = a.dtypes
+            index_value = parse_index(pd.RangeIndex(a.shape[0]))
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: Need to consider `a` has unknown shape on axis 0.

File: mars/dataframe/base/reset_index.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        else:
+            empty_df = build_empty_df(a.dtypes)
+            empty_df.index = a.index_value.to_pandas()
+            empty_df = empty_df.reset_index(level=self.level, col_level=self.col_level, col_fill=self.col_fill)
+            shape = (a.shape[0], len(empty_df.columns))
+            columns_value = parse_index(empty_df.columns)
+            dtypes = empty_df.dtypes
+            if isinstance(empty_df.index, pd.RangeIndex):
+                index_value = parse_index(pd.RangeIndex(a.shape[0]))
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: Ditto.

File: mars/dataframe/base/reset_index.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            else:
+                index_value = parse_index(empty_df.index)
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: We'd better to generate `key` for `index_value` and make a util to do that.

File: mars/dataframe/base/reset_index.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        if isinstance(out.index_value.value, IndexValue.RangeIndex):
+            r.index = out.index_value.to_pandas()
+        ctx[out.key] = r
+
+    def _call_series(self, a):
+        if self.drop:
+            index_value = parse_index(pd.RangeIndex(a.shape[0]))
+            return self.new_series([a], shape=a.shape, dtype=a.dtype, name=a.name, index_value=index_value)
+        else:
+            empty_df = pd.Series(index=a.index_value.to_pandas(), dtype=a.dtype, name=a.name)
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: Rename `empty_df` to `empty_series` and make a utils for `build_empty_series` just like `build_empty_df`?

File: mars/dataframe/base/reset_index.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    def _tile_dataframe(cls, op):
+        in_df = op.inputs[0]
+        out_df = op.outputs[0]
+        added_columns_num = len(out_df.dtypes) - len(in_df.dtypes)
+        out_chunks = []
+        range_index = isinstance(out_df.index_value.value, IndexValue.RangeIndex)
+        cum_range = np.cumsum((0, ) + in_df.nsplits[0])
+        for c in in_df.chunks:
+            if range_index:
+                index_value = parse_index(pd.RangeIndex(cum_range[c.index[0]], cum_range[c.index[0] + 1]))
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: Need to consider nan shape as well.

File: mars/dataframe/base/reset_index.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+    @classmethod
+    def _tile_series(cls, op):
+        out_chunks = []
+        out = op.outputs[0]
+        range_index = isinstance(out.index_value.value, IndexValue.RangeIndex)
+        cum_range = np.cumsum((0, ) + op.inputs[0].nsplits[0])
+        for c in op.inputs[0].chunks:
+            if range_index:
+                index_value = parse_index(pd.RangeIndex(cum_range[c.index[0]], cum_range[c.index[0] + 1]))
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: Ditto.

File: mars/dataframe/base/tests/test_base.py [hunk 8]
```diff
@@ -155,3 +155,57 @@ def testRechunk(self):
         series = get_tiled(series)
         self.assertEqual(series2.chunk_shape, series.chunk_shape)
         self.assertEqual(series2.nsplits, series.nsplits)
+
+    def testResetIndex(self):
```
Comment (qinxuye) [inline mars/dataframe/base/tests/test_base.py]: `col_level` and `col_fill` not tested.

File: mars/dataframe/base/reset_index.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    _op_type_ = OperandDef.RESET_INDEX
+
+    _level = AnyField('level')
+    _drop = BoolField('drop')
+    _name = StringField('name')
+    _col_level = AnyField('col_level')
+    _col_fill = AnyField('col_fill')
+
+    def __init__(self, level=None, drop=None, name=None, col_level=None, col_fill=None, object_type=None, **kwargs):
+        super(DataFrameResetIndex, self).__init__(_level=level, _drop=drop, _name=name, _col_level=col_level,
```
Comment (wjsi) [inline mars/dataframe/base/reset_index.py]: super() now support calling without args in Python 3.

File: mars/core.py [hunk 10]
```diff
@@ -61,11 +61,17 @@ def _values_(self):
...
+            return self._key
+        else:
+            return (type(self), *self._values_)
+
     def _obj_set(self, k, v):
         object.__setattr__(self, k, v)
 
     def _update_key(self):
-        self._obj_set('_key', tokenize(type(self), *self._values_))
+        self._obj_set('_key', tokenize(type(self).__name__, *self._values_))
```
Comment (qinxuye) [inline mars/core.py]: Why modify this line? I prefer type actually.

File: mars/core.py [hunk 11]
```diff
@@ -244,7 +253,7 @@ def device(self):
 
     def _update_key(self):
         object.__setattr__(self, '_key', tokenize(
-            type(self), *(getattr(self, k, None) for k in self._keys_ if k != '_index')))
+            type(self).__name__, *(getattr(self, k, None) for k in self._keys_ if k != '_index')))
```
Comment (qinxuye) [inline mars/core.py]: Ditto.

File: mars/dataframe/base/reset_index.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        columns_splits = list(in_df.nsplits[1])
+        columns_splits[0] += added_columns_num
+        nsplits = (in_df.nsplits[0], tuple(columns_splits))
+        return new_op.new_dataframes(op.inputs, out_df.shape, nsplits=nsplits,
+                                     chunks=out_chunks, dtypes=out_df.dtypes,
+                                     index_value=out_df.index_value, columns_value=out_df.columns_value)
+
+    @classmethod
+    def tile(cls, op):
+        if op.inputs[0].op.object_type == ObjectType.dataframe:
```
Comment (qinxuye) [inline mars/dataframe/base/reset_index.py]: If the condition looks so tedious,  I'd prefer the previous isinstance one.

File: mars/tensor/random/core.py [hunk 13]
```diff
@@ -310,7 +310,7 @@ def args(self):
                 if slot not in set(TensorRandomOperand.__slots__)]
 
     def _update_key(self):
-        self._key = tokenize(type(self),
+        self._key = tokenize(type(self).__name__,
```
Comment (qinxuye) [inline mars/tensor/random/core.py]: Like I said,  I prefer type because I don't know if name is enough.

**Pull request merged**